internal/cryptokit/hmac.go

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//go:build darwin
+
+package cryptokit
+
+// #include "cryptokit.h"
+import "C"
+import (
+	"errors"
+	"hash"
+	"runtime"
+	"unsafe"
+)
+
+var _ hash.Hash = (*cryptoKitHMAC)(nil)
+var _ cloneHash = (*cryptoKitHMAC)(nil)
+
+type cryptoKitHMAC struct {
+	pinner runtime.Pinner
+	ptr    unsafe.Pointer
+
+	hashEnum int
+	key      []byte
+
+	blockSize int
+	size      int
+}
+
+func NewHMAC(fh func() hash.Hash, key []byte) hash.Hash {
+	h := fh()
+	if h == nil {
+		return nil
+	}
+
+	hashEnum := hashToHMACEnum(h)
+	if hashEnum == 0 {
+		// The hash function is not supported by the HMAC implementation.
+		return nil
+	}
+
+	pinner := runtime.Pinner{}
+
+	if len(key) > 0 {
+		pinner.Pin(&key[0])
+		defer pinner.Unpin()
+	}
+
+	hmac := &cryptoKitHMAC{
+		pinner: pinner,
+		ptr: C.initMAC(
+			C.int(hashEnum),
+			base(key), C.int(len(key)),
+		),
+		hashEnum:  hashEnum,
+		key:       key,
+		blockSize: h.BlockSize(),
+		size:      h.Size(),
+	}
+
+	runtime.SetFinalizer(hmac, func(h *cryptoKitHMAC) {
+		if h.ptr != nil {
+			C.freeHMAC(
+				C.int(h.hashEnum),
+				h.ptr,
+			)
+		}
+	})
+
+	return hmac
+}
+
+func (h *cryptoKitHMAC) Write(p []byte) (n int, err error) {
+	if len(p) > 0 {
+		h.pinner.Pin(&p[0])
+		defer h.pinner.Unpin()
+	}
+
+	C.updateHMAC(C.int(h.hashEnum),
+		h.ptr,
+		base(p), C.int(len(p)))
+
+	runtime.KeepAlive(h)
+
+	return len(p), nil
+}
+
+func (h *cryptoKitHMAC) Sum(b []byte) []byte {
+	hashSlice := make([]byte, h.size, 64) // explicit cap to allow stack allocation
+	C.finalizeHMAC(
+		C.int(h.hashEnum),
+		h.ptr,
+		base(hashSlice),
+	)
+	runtime.KeepAlive(h)
+
+	b = append(b, hashSlice...)
+
+	return b
+}
+
+func (h *cryptoKitHMAC) MarshalBinary() ([]byte, error) {
+	return nil, errors.New("cryptokit: hash state is not marshallable")
+}
+
+func (h *cryptoKitHMAC) AppendBinary(b []byte) ([]byte, error) {
+	return nil, errors.New("cryptokit: hash state is not marshallable")
+}
+
+func (h *cryptoKitHMAC) UnmarshalBinary(data []byte) error {
+	return errors.New("cryptokit: hash state is not marshallable")
+}
+
+func (h *cryptoKitHMAC) Clone() hash.Hash {
+	panic("cryptokit: hash state is not cloneable")
+}
```

Comment:
The HMAC type don't need to implement this methods. Please remove.

Reply:
it is, because it is returning hash.Hash in standard library. And in 1.25, if we don't return this types in interface, it will give an error.

Reply:
https://pkg.go.dev/crypto/hmac
please check